File: src/xenly.c
/*
 * XENLY - high-level and general-purpose programming language
 * created, designed, and developed by Cyril John Magayaga (cjmagayaga957@gmail.com, cyrilmagayaga@proton.me).
 *
 * It is initially written in C programming language.
 *
 * It is available for Linux and Windows operating systems.
 *
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <stdbool.h>
#include "xenly.h"

// Array to store all declared variables
Variable variables[MAX_VARIABLES];
int variable_count = 0;
bool in_multi_line_comment = false;

void process_line(char* line) {
    char* processed_line = remove_comments(line);
    if (processed_line && *processed_line) {
        interpret_line(processed_line);
    }
}

char* remove_comments(char* line) {
    static char cleaned_line[MAX_LINE_LENGTH];
    char* write = cleaned_line;
    char* read = line;
    
    while (*read) {
        if (in_multi_line_comment) {
            if (read[0] == '*' && read[1] == '/') {
                in_multi_line_comment = false;
                read += 2;
            }
            
            else {
                read++;
            }
        }
        
        else if (read[0] == '/' && read[1] == '/') {
            break;  // Ignore rest of the line
        }
        
        else if (read[0] == '/' && read[1] == '*') {
            in_multi_line_comment = true;
            read += 2;
        }
        
        else {
            *write++ = *read++;
        }
    }
    *write = '\0';
    
    return cleaned_line[0] ? cleaned_line : NULL;
}

void interpret_line(char* line) {
    char* trimmed_line = trim(line);
    
    if (strncmp(trimmed_line, "nota(", 5) == 0) {
        char* args = trimmed_line + 5;
        char* end = strrchr(args, ')');
        if (end) *end = '\0';
        print_function(args);
    }
    
    else if (strncmp(trimmed_line, "var ", 4) == 0) {
        var_declaration(trimmed_line + 4);
    }
    
    else if (strncmp(trimmed_line, "bool ", 5) == 0) {
        bool_declaration(trimmed_line + 5);
    }
}

void print_function(char* args) {
    char* trimmed_args = trim(args);
    
    if (trimmed_args[0] == '"' && trimmed_args[strlen(trimmed_args)-1] == '"') {
        // String literal
        trimmed_args[strlen(trimmed_args)-1] = '\0';
        printf("%s\n", trimmed_args + 1);
    }
    
    else {
        // Variable or expression
        char* value = get_variable_value(trimmed_args);
        if (value != NULL) {
            printf("%s\n", value);
        }
        
        else {
            // Assume it's a numeric expression
            double result = evaluate_expression(trimmed_args);
            printf("%g\n", result);
        }
    }
}

void var_declaration(char* args) {
    char* name = strtok(args, "=");
    char* value = strtok(NULL, "");
    
    if (!name || !value) {
        printf("Error: Invalid variable declaration\n");
        return;
    }
    
    name = trim(name);
    value = trim(value);

    if (variable_count >= MAX_VARIABLES) {
        printf("Error: Maximum number of variables reached.\n");
        return;
    }

    strcpy(variables[variable_count].name, name);

    if (value[0] == '"' && value[strlen(value)-1] == '"') {
        variables[variable_count].type = 's';
        strncpy(variables[variable_count].value.s, value + 1, strlen(value) - 2);
        variables[variable_count].value.s[strlen(value) - 2] = '\0';
    }
    
    else if (strchr(value, '.') != NULL) {
        variables[variable_count].type = 'f';
        variables[variable_count].value.f = atof(value);
    }
    
    else {
        variables[variable_count].type = 'i';
        variables[variable_count].value.i = atoi(value);
    }

    variable_count++;
}

void bool_declaration(char* args) {
    char* name = strtok(args, "=");
    char* value = strtok(NULL, "");
    
    if (!name || !value) {
        printf("Error: Invalid boolean declaration\n");
        return;
    }
    
    name = trim(name);
    value = trim(value);

    if (variable_count >= MAX_VARIABLES) {
        printf("Error: Maximum number of variables reached.\n");
        return;
    }

    strcpy(variables[variable_count].name, name);
    variables[variable_count].type = 'b';
    variables[variable_count].value.b = parse_bool(value);

    variable_count++;
}

char* get_variable_value(const char* name) {
    for (int i = 0; i < variable_count; i++) {
        if (strcmp(variables[i].name, name) == 0) {
            static char buffer[100];
            switch (variables[i].type) {
                case 's':
                    return variables[i].value.s;
                
                case 'i':
                    sprintf(buffer, "%d", variables[i].value.i);
                    return buffer;
                
                case 'f':
                    sprintf(buffer, "%g", variables[i].value.f);
                    return buffer;
                
                case 'b':
                    sprintf(buffer, "%d", variables[i].value.b);
                    return buffer;
            }
        }
    }
    return NULL;
}

double evaluate_expression(const char* expr) {
    char expr_copy[MAX_LINE_LENGTH];
    strcpy(expr_copy, expr);
    
    char* token = strtok(expr_copy, " ");
    double result = atof(token);
    
    while ((token = strtok(NULL, " ")) != NULL) {
        char op = token[0];
        token = strtok(NULL, " ");
        double num = atof(token);
        
        switch (op) {
            case '+': result += num; break;
            case '-': result -= num; break;
            case '*': result *= num; break;
            case '/': result /= num; break;
        }
    }
    
    return result;
}

char* trim(char* str) {
    while(isspace((unsigned char)*str)) str++;
    if(*str == 0) return str;
    char* end = str + strlen(str) - 1;
    while(end > str && isspace((unsigned char)*end)) end--;
    end[1] = '\0';
    return str;
}

bool parse_bool(const char* value) {
    if (strcmp(value, "true") == 0) {
        return true;
    }
    
    else if (strcmp(value, "false") == 0) {
        return false;
    }
    
    else {
        printf("Error: Invalid boolean value '%s'. Defaulting to false.\n", value);
        return false;
    }
}

File: src/xenly.rs
/*
 * XENLY - high-level and general-purpose programming language
 * created, designed, and developed by Cyril John Magayaga (cjmagayaga957@gmail.com, cyrilmagayaga@proton.me).
 *
 * It is initially written in Rust programming language.
 * 
 * It is available for Linux and Windows operating systems.
 *
 */
use std::str::FromStr;
use std::collections::HashMap;

const MAX_VARIABLES: usize = 100;

#[derive(Debug, Clone)]
enum VariableValue {
    String(String),
    Integer(i32),
    Float(f32),
    Boolean(bool),
}

#[derive(Debug, Clone)]
struct Variable {
    name: String,
    value: VariableValue,
}

struct Interpreter {
    variables: HashMap<String, Variable>,
    in_multi_line_comment: bool,
}

impl Interpreter {
    fn new() -> Self {
        Self {
            variables: HashMap::new(),
            in_multi_line_comment: false,
        }
    }

    fn process_line(&mut self, line: &str) {
        if let Some(processed_line) = self.remove_comments(line) {
            self.interpret_line(&processed_line);
        }
    }

    fn remove_comments(&mut self, line: &str) -> Option<String> {
        let mut cleaned_line = String::new();
        let mut chars = line.chars().peekable();

        while let Some(&ch) = chars.peek() {
            if self.in_multi_line_comment {
                if ch == '*' && chars.clone().nth(1) == Some('/') {
                    self.in_multi_line_comment = false;
                    chars.next();
                    chars.next();
                }
                
                else {
                    chars.next();
                }
            }
            
            else if ch == '/' && chars.clone().nth(1) == Some('/') {
                break;  // Ignore rest of the line
            }
            
            else if ch == '/' && chars.clone().nth(1) == Some('*') {
                self.in_multi_line_comment = true;
                chars.next();
                chars.next();
            }
            
            else {
                cleaned_line.push(ch);
                chars.next();
            }
        }

        if cleaned_line.is_empty() {
            None
        } else {
            Some(cleaned_line)
        }
    }

    fn interpret_line(&mut self, line: &str) {
        let trimmed_line = line.trim();

        if trimmed_line.starts_with("nota(") {
            let args = &trimmed_line[5..trimmed_line.len() - 1]; // Remove "nota(" and ")"
            self.print_function(args);
        }
        
        else if trimmed_line.starts_with("var ") {
            self.var_declaration(&trimmed_line[4..]);
        }
        
        else if trimmed_line.starts_with("bool ") {
            self.bool_declaration(&trimmed_line[5..]);
        }
    }

    fn print_function(&self, args: &str) {
        let trimmed_args = args.trim();

        if trimmed_args.starts_with('"') && trimmed_args.ends_with('"') {
            // String literal
            println!("{}", &trimmed_args[1..trimmed_args.len() - 1]);
        }
        
        else if let Some(value) = self.get_variable_value(trimmed_args) {
            // Variable
            println!("{}", value);
        }
        
        else {
            // Assume it's a numeric expression
            let result = self.evaluate_expression(trimmed_args);
            println!("{}", result);
        }
    }

    fn var_declaration(&mut self, args: &str) {
        let parts: Vec<&str> = args.split('=').collect();
        if parts.len() != 2 {
            eprintln!("Error: Invalid variable declaration");
            return;
        }

        let name = parts[0].trim();
        let value = parts[1].trim();

        if self.variables.len() >= MAX_VARIABLES {
            eprintln!("Error: Maximum number of variables reached.");
            return;
        }

        let variable = if value.starts_with('"') && value.ends_with('"') {
            Variable {
                name: name.to_string(),
                value: VariableValue::String(value[1..value.len() - 1].to_string()),
            }
        }
        
        else if value.contains('.') {
            Variable {
                name: name.to_string(),
                value: VariableValue::Float(f32::from_str(value).unwrap()),
            }
        }
        
        else {
            Variable {
                name: name.to_string(),
                value: VariableValue::Integer(i32::from_str(value).unwrap()),
            }
        };

        self.variables.insert(name.to_string(), variable);
    }

    fn bool_declaration(&mut self, args: &str) {
        let parts: Vec<&str> = args.split('=').collect();
        if parts.len() != 2 {
            eprintln!("Error: Invalid boolean declaration");
            return;
        }

        let name = parts[0].trim();
        let value = parts[1].trim();

        if self.variables.len() >= MAX_VARIABLES {
            eprintln!("Error: Maximum number of variables reached.");
            return;
        }

        let variable = Variable {
            name: name.to_string(),
            value: VariableValue::Boolean(self.parse_bool(value)),
        };

        self.variables.insert(name.to_string(), variable);
    }

    fn get_variable_value(&self, name: &str) -> Option<String> {
        self.variables.get(name).map(|var| match &var.value {
            VariableValue::String(s) => s.clone(),
            VariableValue::Integer(i) => i.to_string(),
            VariableValue::Float(f) => f.to_string(),
            VariableValue::Boolean(b) => b.to_string(),
        })
    }

    fn evaluate_expression(&self, expr: &str) -> f64 {
        let mut result = 0.0;
        let mut tokens = expr.split_whitespace();

        if let Some(first) = tokens.next() {
            result = f64::from_str(first).unwrap_or(0.0);
        }

        while let Some(op) = tokens.next() {
            if let Some(num) = tokens.next() {
                let num = f64::from_str(num).unwrap_or(0.0);
                match op {
                    "+" => result += num,
                    "-" => result -= num,
                    "*" => result *= num,
                    "/" => result /= num,
                    _ => (),
                }
            }
        }

        result
    }

    fn parse_bool(&self, value: &str) -> bool {
        match value {
            "true" => true,
            "false" => false,
            _ => {
                eprintln!("Error: Invalid boolean value '{}'. Defaulting to false.", value);
                false
            }
        }
    }
}

pub fn process_line(line: &str) {
    static mut INTERPRETER: Option<Interpreter> = None;
    unsafe {
        if INTERPRETER.is_none() {
            INTERPRETER = Some(Interpreter::new());
        }

        if let Some(interpreter) = &mut INTERPRETER {
            interpreter.process_line(line);
        }
    }
}
